### PR TITLE
Check variable assignments

### PR DESCRIPTION
**Pull request by agronholm**
This type checks assignments to annotated local variables and arguments in functions.

### CODE REVIEW

**Comment by agronholm** [comment]
This PR is now a mish-mash of the feature in question, plus all the changes I've made since then. Sorry about that.

**Comment by agronholm** [comment]
Found a bug: PEP 604 unions in variable annotations are not transformed into `typing.Union` on Python < 3.10. Fixing.

File: src/typeguard/_functions.py [hunk 1]
```diff
@@ -260,6 +260,25 @@ def check_yield_type(yieldval: T, memo: CallMemo) -> T:
     return yieldval
 
 
+def check_variable_assignment(
+    value: object, expected_annotations: dict[str, Any], memo: CallMemo
```
Comment (marcelroed) [inline src/typeguard/_functions.py:264]: Why are you using the object type instead of Any?

File: src/typeguard/_transformer.py [hunk 2]
```diff
@@ -429,3 +478,105 @@ def visit_Yield(self, node: Yield):
...
+            )
+            expected_types = Dict(keys=[Constant(node.target.id)], values=[annotation])
+            check_call = Call(
+                self._get_typeguard_import("check_variable_assignment"),
+                [operator_call, expected_types, self._memo.get_call_memo_name()],
+                [],
+            )
+            node = Assign(targets=[node.target], value=check_call)
+
+        return node
```
Comment (marcelroed) [inline src/typeguard/_transformer.py:582]: Looks good, but I would like docstrings to explain the changes on visit for readability, perhaps showing what transformation is made with syntax

File: src/typeguard/_functions.py (hunk 1 above)
Comment (agronholm) [reply src/typeguard/_functions.py:264]: To treat `value` as opaque. Making it an `object` makes access to any of its (non-dunder) attribute invalid. I probably should've done this in many other places too.

**Comment by agronholm** [comment]
> Found a bug: PEP 604 unions in variable annotations are not transformed into `typing.Union` on Python < 3.10. Fixing.

Fixed.

File: src/typeguard/_transformer.py (hunk 2 above)
Comment (agronholm) [reply src/typeguard/_transformer.py]: I've added some docstrings to methods that make changes to the nodes passing through them. Is this satisfactory or would you say it needs more?

Comment (marcelroed) [reply src/typeguard/_transformer.py]: Looks good! I'm certainly satisfied.

**Comment by agronholm** [comment]
Ok, one step closer now :smiley:
I'm going to make another change that avoids checking names imported within an `if TYPE_CHECKING:` branch.

**Pull request merged**